Explore: Redux migration

davkal opened the pull request:
Migrate state management from top Explore component to redux.

*Functional changes*

- added resize handler to calculate graph interval
- changed graph interval to be string interval and Millies
- removed timer delay for scanning

*Non-functional changes*

- Moved Explore logic to Action creators and reducer
- Hooked Explore into global redux store, via the `explore` slice
- Moved interval calculation to reducer
- Former Explore state is now `ExploreItemState` (with separate reducer) to allow split views
- Item states are properties of the Global explore state as `left` and `right`
- Connected Explore child components to redux store

*Discussion*

- In general, action creators try not to use state via `getState` to keep them pure, with some exceptions
- Main logic remains in action creators; reducers should only put results into state, with minimal computation
- The original implementation had no tests, so I'd argue we continue with this PR without tests as well, and implement them in a separate PR.

davkal committed:
WIP Explore redux migration

torkelo commented on public/app/features/explore/Explore.tsx:
```diff
@@ -1035,4 +542,61 @@ export class Explore extends React.PureComponent<ExploreProps, ExploreState> {
+  highlightLogsExpression,
+  initializeExplore,
+  modifyQueries,
+  onSize: changeSize, // used by withSize HOC
```
I would like to stop using size-me, for the react panels I am using AutoSizer from react-virtualized (https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/dashgrid/PanelChrome.tsx#L91) 
It uses render props instead and has been more reliable (withSize/sizeMe) has some strange issues that were solved by switching to AutoSizer (which I think i using a more optimized approach). But need create a tech/chore task to remove all other uses of withSize/size-me.  Think it's only DashboardGrid, explore graph, and now here. 

btw why does the explore root component care about size?

dprokop replied:
@davkal can you elaborate a little bit more what is the relation between graph's width and interval?

davkal committed:
Allow multiple Explore items for split

davkal committed:
Save state in URL and fix tests

davkal committed:
Move types to types/explore

davkal committed:
Update comments

davkal committed:
Connect Explore child components to store

davkal committed:
Fix reducer issues

dprokop commented on public/app/features/explore/state/actions.ts:
```diff
@@ -0,0 +1,836 @@
+  StateSave = 'STATE_SAVE',
+}
+
+export interface AddQueryRowAction {
```
As there is a lot of keystrokes required to define those interfaces and action creators. Have you experimented with `redux-actions` or `typesafe-actions` @davkal?

davkal replied:
I have not. My main goal was to be consistent with the rest of the codebase. This was not the time to try new things. But if you have had good experience with them, perhaps we can migrate all redux actions in a separate endeavour.

dprokop replied:
👍

dprokop commented on public/app/features/explore/state/actions.ts:
```diff
@@ -0,0 +1,836 @@
+  type: ActionTypes.StateSave;
+}
+
+export type Action =
```
WDYT about moving the action types to a separate file to make this one to contain action creators and thunks only?  The other thing that comes to my mind is to split this file into actions that would be grouped by concern, e.g. UI, datasource, query...

davkal replied:
Moving action types to a file sounds good. I'm less convinced about grouping, my fear is that it will result in uncertainty where something is if the groups are hazy.

dprokop replied:
Yes, I agree. There is no clear separation which action should belong to which concern. Ignore that part of a comment :)

dprokop commented on public/app/features/explore/state/actions.ts:
```diff
@@ -0,0 +1,836 @@
+  };
+}
+
+export function clickClear(exploreId: ExploreId): ThunkResult<void> {
```
Talking about those actions that names start with `click...`. IMO they introduce an unnecessary level of indirectness into Explore's code. Take a look at the usage: 

```ts
 onClickClear = () => {
    this.props.clickClear(this.props.exploreId);
};
```

When I read this code the only thing I learn is that someone clicked clear. What I would like to learn though is what is the effect of the click, in this case this would probably be `clearAllQueries` or `clearQueriesAndPersistState` depending on how much information is really necessary.

davkal replied:
IMO handlers, actions and reducers are 3 concerns. But sometimes they overlap and this is where it gets messy. Reducers should be dumb, but not as dumb as to blindly apply only the state that they have been given (otherwise, what are they reducing?). Handler should translate to an action, because that's what enters the state logic. If the actions are too abstracted, they prescribe the state change (IMO they should not). Instead they should be reusable from an action point of view: "clear all" may be a button that occurs on the top bar, but also in a menu, etc.
For that concrete case I agree that the better name would be `clearAll`. But I'd resist on saying what it clears.

torkelo replied:
I agree with some of this, the click part of the action names & function names seems unnecessary. Would prefer to have it be more domain oriented. CLEAR_EXPLORE_QUERIES. 
CLOSE_EXPLORE_SPLIT, or prefix the actions name constants with EXPLORE.

A lot of these action names needs to be changed to include EXPLORE as they are too generic and action names are global.

dprokop reviewed (commented):
Quite huge one... Left couple of comments, though;)

davkal commented on public/app/features/explore/state/reducers.ts:
```diff
@@ -0,0 +1,437 @@
+/**
+ * Reducer for an Explore area, to be used by the global Explore reducer.
+ */
+const itemReducer = (state, action: Action): ExploreItemState => {
```
I followed the advice in the [docs](https://redux.js.org/recipes/structuring-reducers/splitting-reducer-logic)
Perhaps you can help me find an obvious slice for more composition?
As for splitting up by grouping thematically, I think this is more confusing IMO and AFAIK I dont think it affects testability.
Also, since reading Ousterhout, I'm less inclined to split things up for organisational reasons, but perhaps I'm overlooking other benefits.

dprokop replied:
Maybe simply defining a simple reducer per switch statement would be sufficient. So, instead of having:

```ts
case ActionTypes.ChangeTime: {
  return {
     ...state,
    range: action.range,
  };
}
```

I would go with:
```
case ActionTypes.ChangeTime: changeTimeReducer(state, actionPayload)
```
 
This way we could have reducers more testable.

davkal committed:
File organization, action naming, comments

- moved ActionTypes to `./state/actionTypes`
- renamed click-related actions
- added comments to actions and state types
- prefixed Explore actions with `explore/`
- fixed query override issue when row was added

torkelo commented on public/app/features/explore/state/actionTypes.ts:
```diff
@@ -0,0 +1,252 @@
+export interface ModifyQueriesAction {
+  type: ActionTypes.ModifyQueries;
+  exploreId: ExploreId;
+  modification: any;
```
string?

davkal committed:
Move action properties to payload
